pkg/filesystem/path/resolve.go

```diff
@@ -1,127 +1,60 @@
 package path
 
-import (
-	"strings"
-
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
-)
-
-func stripOneOrMoreSlashes(p string) string {
-	for {
-		p = p[1:]
-		if p == "" || p[0] != '/' {
-			return p
-		}
-	}
-}
-
 type resolverState struct {
-	stack           []string
+	stack           []RelativeParser
 	componentWalker ComponentWalker
 }
 
-// Push a new path string onto the stack of paths that need to be
+// Push a new path onto the stack of paths that need to be
 // processed. This happens once when the resolution process starts, and
 // will happen for every symlink encountered.
-func (rs *resolverState) push(scopeWalker ScopeWalker, path string) error {
-	// Unix-style paths are generally passed to system calls that
-	// accept C strings. There is no way these can accept null
-	// bytes.
-	if strings.ContainsRune(path, '\x00') {
-		return status.Error(codes.InvalidArgument, "Path contains a null byte")
-	}
-
-	// Determine whether the path is absolute.
-	absolute := false
-	if path != "" && path[0] == '/' {
-		path = stripOneOrMoreSlashes(path)
-		absolute = true
-	}
-
+func (rs *resolverState) push(scopeWalker ScopeWalker, parser Parser) error {
 	// Push the path without any leading slashes onto the stack, so
-	// that its components may be processed. Apply  them against the
+	// that its components may be processed. Apply them against the
 	// right directory.
-	componentWalker, err := scopeWalker.OnScope(absolute)
+	componentWalker, remainder, err := parser.ParseScope(scopeWalker)
 	if err != nil {
 		return err
 	}
-	rs.stack = append(rs.stack, path)
+	if remainder != nil {
+		rs.stack = append(rs.stack, remainder)
+	}
 	rs.componentWalker = componentWalker
 	return nil
 }
 
 // Pop a single filename from the stack of paths that need to be
 // processed. This filename is the first one that should be processed.
-func (rs *resolverState) pop() string {
+func (rs *resolverState) pop() (GotDirectoryOrSymlink, error) {
 	p := &rs.stack[len(rs.stack)-1]
```

Comment:
Considering that `rs.stack` is now a stack of interface objects, you can drop the `&`. This will make the call to `ParseFirstComponent()` cleaner.

Alternatively, just drop `p` entirely and inline it?

Reply:
I would prefer not to inline it, the next statement would become an eyesore.